stlab/forest.hpp

```diff
@@ -0,0 +1,1240 @@
+/*
+    Copyright 2013 Adobe
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+/**************************************************************************************************/
+
+#ifndef STLAB_FOREST_HPP
+#define STLAB_FOREST_HPP
+
+/**************************************************************************************************/
+
+#include <cassert>
+#include <cstddef>
+#include <functional>
+#include <iterator>
+
+#include <stlab/algorithm/reverse.hpp>
+#include <stlab/iterator/set_next.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+enum class forest_edge : bool { trailing, leading };
+
+constexpr auto forest_trailing_edge{forest_edge::trailing};
+constexpr auto forest_leading_edge{forest_edge::leading};
+
+/**************************************************************************************************/
+
+inline auto pivot(forest_edge e) { return forest_edge(static_cast<bool>(e) ^ 1); }
+
+template <typename I> // I models FullorderIterator
+inline void pivot(I& i) {
```

Comment:
I think we should drop the use of inline for template functions. I think it will be ignored anyway.

Reply:
Done